File: apple/CaCompteKit/Sources/Domain/Rules/GameDefinition.swift
import Foundation

/// Définition déclarative d'un jeu — transposition mécanique de
/// `spec/schema/game-definition.schema.json`. Couvre tout ce qui est exprimable en données ;
/// le calcul non trivial est délégué au moteur impératif désigné par `engine`.
public struct GameDefinition: Identifiable, Sendable, Codable, Equatable {
  public struct LocalizedText: Sendable, Codable, Equatable {
    public let fr: String
    public let en: String?
    public let es: String?
    public let de: String?
    public let it: String?

    public init(
      fr: String, en: String? = nil, es: String? = nil, de: String? = nil, it: String? = nil
    ) {
      self.fr = fr
      self.en = en
      self.es = es
      self.de = de
      self.it = it
    }

    /// Doc utilisateur — résolution de présentation, pas de calcul : ne porte sur aucune valeur
    /// que rejoue un golden file (qui ne vérifie que les nombres du domaine), donc ne remet pas
    /// en cause le déterminisme protégé par ADR-0002 malgré la lecture de `Bundle.main` ici.
    /// `Bundle.main.preferredLocalizations` respecte à la fois les langues déclarées par l'app
    /// (`knownRegions`) et le réglage de langue par app d'iOS (Réglages > Ça Compte > Langue),
    /// contrairement à `Locale.current` qui ignore ce réglage par app. Repli sur le français —
    /// la langue source, toujours renseignée — si la traduction demandée est absente.
    public var localized: String {
      switch Bundle.main.preferredLocalizations.first {
      case "en": en ?? fr
      case "es": es ?? fr
      case "de": de ?? fr
      case "it": it ?? fr
      default: fr
      }
    }

    /// Doc utilisateur — remontée : la recherche de jeux ne doit pas dépendre de la langue
    /// affichée par l'app. Contrairement à `localized` (une seule traduction, celle à
    /// afficher), ceci compare `term` à *toutes* les traductions déclarées (fr toujours
    /// présente, les autres si fournies) — un jeu reste trouvable même si son nom cherché
    /// correspond à une langue différente de celle actuellement affichée.
    ///
    /// Doc utilisateur — remontée : les espaces (y compris internes, pas seulement en début/fin)
    /// ne doivent pas compter — « petit bac », « petitbac » et « petit  bac » doivent tous
    /// trouver le même jeu. Les deux côtés de la comparaison sont donc dépouillés de leurs
    /// espaces avant le `contains`.
    public func matches(_ term: String) -> Bool {
      let strippedTerm = term.filter { !$0.isWhitespace }
      for candidate in [fr, en, es, de, it].compactMap({ $0 }) {
        if candidate.filter({ !$0.isWhitespace }).localizedCaseInsensitiveContains(strippedTerm) {
          return true
        }
      }
      return false
    }
  }

  public enum PaletteToken: String, Sendable, Codable, Equatable {
    case ink, brass, teal, azur, ambre, emeraude, magenta, ardoise, cyan, vermillon, violet, olive
  }

  public struct Players: Sendable, Codable, Equatable {
    public struct Teams: Sendable, Codable, Equatable {
      public let size: Int
      public let fixed: Bool

      private enum CodingKeys: String, CodingKey { case size, fixed }

      public init(size: Int, fixed: Bool = true) {
        self.size = size
        self.fixed = fixed
      }

      public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        size = try container.decode(Int.self, forKey: .size)
        fixed = try container.decodeIfPresent(Bool.self, forKey: .fixed) ?? true
      }
    }

    public let min: Int
    public let max: Int
    public let recommended: [Int]
    public let teams: Teams?

    private enum CodingKeys: String, CodingKey { case min, max, recommended, teams }

    public init(min: Int, max: Int, recommended: [Int] = [], teams: Teams? = nil) {
      self.min = min
      self.max = max
      self.recommended = recommended
      self.teams = teams
    }

    public init(from decoder: Decoder) throws {
      let container = try decoder.container(keyedBy: CodingKeys.self)
      min = try container.decode(Int.self, forKey: .min)
      max = try container.decode(Int.self, forKey: .max)
      recommended = try container.decodeIfPresent([Int].self, forKey: .recommended) ?? []
      teams = try container.decodeIfPresent(Teams.self, forKey: .teams)
    }
  }

  public enum EntryKind: String, Sendable, Codable, Equatable {
    case integer, categorySheet, structured, rank, predictionAndResult
  }

  public struct Category: Sendable, Codable, Equatable {
    public enum Section: String, Sendable, Codable, Equatable { case upper, lower }

    public struct Scoring: Sendable, Codable, Equatable {
      public enum Kind: String, Sendable, Codable, Equatable { case fixed, sumOfDice, multipleOf }
      public let kind: Kind
      public let value: Int?
      public let max: Int?

      public init(kind: Kind, value: Int? = nil, max: Int? = nil) {
        self.kind = kind
        self.value = value
        self.max = max
      }
    }

    public let id: String
    public let label: LocalizedText
    public let section: Section
    public let scoring: Scoring

    public init(id: String, label: LocalizedText, section: Section, scoring: Scoring) {
      self.id = id
      self.label = label
      self.section = section
      self.scoring = scoring
    }
  }

  public struct RankOption: Sendable, Codable, Equatable {
    public let id: String
    public let label: LocalizedText
    public let points: Int

    public init(id: String, label: LocalizedText, points: Int) {
      self.id = id
      self.label = label
      self.points = points
    }
  }

  public struct ScoreEntrySpec: Sendable, Codable, Equatable {
    public let kind: EntryKind
    public let min: Int?
    public let max: Int?
    public let step: Int
    public let allowsNegative: Bool
    public let warnBelow: Int?
    public let warnAbove: Int?
    public let categories: [Category]?
    public let ranks: [RankOption]?

    private enum CodingKeys: String, CodingKey {
      case kind, min, max, step, allowsNegative, warnBelow, warnAbove, categories, ranks
    }

    public init(
      kind: EntryKind,
      min: Int? = nil,
      max: Int? = nil,
      step: Int = 1,
      allowsNegative: Bool = false,
      warnBelow: Int? = nil,
      warnAbove: Int? = nil,
      categories: [Category]? = nil,
      ranks: [RankOption]? = nil
    ) {
      self.kind = kind
      self.min = min
      self.max = max
      self.step = step
      self.allowsNegative = allowsNegative
      self.warnBelow = warnBelow
      self.warnAbove = warnAbove
      self.categories = categories
      self.ranks = ranks
    }

    public init(from decoder: Decoder) throws {
      let container = try decoder.container(keyedBy: CodingKeys.self)
      kind = try container.decode(EntryKind.self, forKey: .kind)
      min = try container.decodeIfPresent(Int.self, forKey: .min)
      max = try container.decodeIfPresent(Int.self, forKey: .max)
      step = try container.decodeIfPresent(Int.self, forKey: .step) ?? 1
      allowsNegative = try container.decodeIfPresent(Bool.self, forKey: .allowsNegative) ?? false
      warnBelow = try container.decodeIfPresent(Int.self, forKey: .warnBelow)
      warnAbove = try container.decodeIfPresent(Int.self, forKey: .warnAbove)
      categories = try container.decodeIfPresent([Category].self, forKey: .categories)
      ranks = try container.decodeIfPresent([RankOption].self, forKey: .ranks)
    }
  }

  public enum ModifierKind: String, Sendable, Codable, Equatable {
    case exclusiveFlag, flag, counter
  }

  public struct Modifier: Sendable, Codable, Equatable {
    public let id: String
    public let label: LocalizedText
    public let kind: ModifierKind
    public let required: Bool
    public let max: Int?

    private enum CodingKeys: String, CodingKey { case id, label, kind, required, max }

    public init(
      id: String, label: LocalizedText, kind: ModifierKind, required: Bool = false, max: Int? = nil
    ) {
      self.id = id
      self.label = label
      self.kind = kind
      self.required = required
      self.max = max
    }

    public init(from decoder: Decoder) throws {
      let container = try decoder.container(keyedBy: CodingKeys.self)
      id = try container.decode(String.self, forKey: .id)
      label = try container.decode(LocalizedText.self, forKey: .label)
      kind = try container.decode(ModifierKind.self, forKey: .kind)
      required = try container.decodeIfPresent(Bool.self, forKey: .required) ?? false
      max = try container.decodeIfPresent(Int.self, forKey: .max)
    }
  }

  public enum Direction: String, Sendable, Codable, Equatable {
    case lowestWins, highestWins, targetExact
  }

  public enum Aggregation: String, Sendable, Codable, Equatable {
    case cumulative, lastRoundOnly, bestRound
  }

  public struct Scoring: Sendable, Codable, Equatable {
    public let direction: Direction
    public let aggregation: Aggregation
    public let roundLabel: LocalizedText?
    public let entry: ScoreEntrySpec
    public let modifiers: [Modifier]

    private enum CodingKeys: String, CodingKey {
      case direction, aggregation, roundLabel, entry, modifiers
    }

    public init(
      direction: Direction,
      aggregation: Aggregation = .cumulative,
      roundLabel: LocalizedText? = nil,
      entry: ScoreEntrySpec,
      modifiers: [Modifier] = []
    ) {
      self.direction = direction
      self.aggregation = aggregation
      self.roundLabel = roundLabel
      self.entry = entry
      self.modifiers = modifiers
    }

    public init(from decoder: Decoder) throws {
      let container = try decoder.container(keyedBy: CodingKeys.self)
      direction = try container.decode(Direction.self, forKey: .direction)
      aggregation =
        try container.decodeIfPresent(Aggregation.self, forKey: .aggregation) ?? .cumulative
      roundLabel = try container.decodeIfPresent(LocalizedText.self, forKey: .roundLabel)
      entry = try container.decode(ScoreEntrySpec.self, forKey: .entry)
      modifiers = try container.decodeIfPresent([Modifier].self, forKey: .modifiers) ?? []
    }
  }

  public enum EndConditionType: String, Sendable, Codable, Equatable {
    case scoreThreshold, roundLimit, targetReached, allSheetsComplete, elimination, manualStop
  }

  public enum Comparison: String, Sendable, Codable, Equatable {
    case greaterThanOrEqual = ">="
    case greaterThan = ">"
    case equalTo = "=="
    case lessThanOrEqual = "<="
    case lessThan = "<"

    public func evaluate(_ lhs: Int, _ rhs: Int) -> Bool {
      switch self {
      case .greaterThanOrEqual: lhs >= rhs
      case .greaterThan: lhs > rhs
      case .equalTo: lhs == rhs
      case .lessThanOrEqual: lhs <= rhs
      case .lessThan: lhs < rhs
      }
    }
  }

  public enum Scope: String, Sendable, Codable, Equatable {
    case anyPlayer, allPlayers, anyTeam
  }

  public struct EndCondition: Sendable, Codable, Equatable {
    public let type: EndConditionType
    public let value: Int?
    public let valueExpression: String?
    public let comparison: Comparison
    public let scope: Scope
    public let completeRound: Bool
    public let variantKey: String?

    private enum CodingKeys: String, CodingKey {
      case type, value, valueExpression, comparison, scope, completeRound, variantKey
    }

    public init(
      type: EndConditionType,
      value: Int? = nil,
      valueExpression: String? = nil,
      comparison: Comparison = .greaterThanOrEqual,
      scope: Scope = .anyPlayer,
      completeRound: Bool = true,
      variantKey: String? = nil
    ) {
      self.type = type
      self.value = value
      self.valueExpression = valueExpression
      self.comparison = comparison
      self.scope = scope
      self.completeRound = completeRound
      self.variantKey = variantKey
    }

    public init(from decoder: Decoder) throws {
      let container = try decoder.container(keyedBy: CodingKeys.self)
      type = try container.decode(EndConditionType.self, forKey: .type)
      value = try container.decodeIfPresent(Int.self, forKey: .value)
      valueExpression = try container.decodeIfPresent(String.self, forKey: .valueExpression)
      comparison =
        try container.decodeIfPresent(Comparison.self, forKey: .comparison) ?? .greaterThanOrEqual
      scope = try container.decodeIfPresent(Scope.self, forKey: .scope) ?? .anyPlayer
      completeRound = try container.decodeIfPresent(Bool.self, forKey: .completeRound) ?? true
      variantKey = try container.decodeIfPresent(String.self, forKey: .variantKey)
    }

    /// `value`, remplacé par la variante choisie si `variantKey` est présent.
    public func resolvedValue(variants: VariantSelection) -> Int {
      if let variantKey {
        return variants.int(variantKey, default: value ?? 0)
      }
      return value ?? 0
    }
  }

  public struct End: Sendable, Codable, Equatable {
    public let conditions: [EndCondition]

    public init(conditions: [EndCondition]) {
      self.conditions = conditions
    }
  }

  public enum TieBreakRule: String, Sendable, Codable, Equatable {
    case bestSingleRound, worstSingleRound, fewestRoundsClosed, mostRoundsWon
    case lowerSecondaryScore, higherSecondaryScore, headToHead, shared
  }

  public enum VariantKind: String, Sendable, Codable, Equatable {
    case bool, integerChoice, integerRange, option
  }

  public struct Variant: Sendable, Codable, Equatable {
    public let id: String
    public let label: LocalizedText
    public let help: LocalizedText?
    public let kind: VariantKind
    public let values: [VariantSelection.Value]
    public let min: Int?
    public let max: Int?
    public let defaultValue: VariantSelection.Value

    private enum CodingKeys: String, CodingKey {
      case id, label, help, kind, values, min, max
      case defaultValue = "default"
    }

    public init(
      id: String,
      label: LocalizedText,
      help: LocalizedText? = nil,
      kind: VariantKind,
      values: [VariantSelection.Value] = [],
      min: Int? = nil,
      max: Int? = nil,
      defaultValue: VariantSelection.Value
    ) {
      self.id = id
      self.label = label
      self.help = help
      self.kind = kind
      self.values = values
      self.min = min
      self.max = max
      self.defaultValue = defaultValue
    }

    public init(from decoder: Decoder) throws {
      let container = try decoder.container(keyedBy: CodingKeys.self)
      id = try container.decode(String.self, forKey: .id)
      label = try container.decode(LocalizedText.self, forKey: .label)
      help = try container.decodeIfPresent(LocalizedText.self, forKey: .help)
      kind = try container.decode(VariantKind.self, forKey: .kind)
      values = try container.decodeIfPresent([VariantSelection.Value].self, forKey: .values) ?? []
      min = try container.decodeIfPresent(Int.self, forKey: .min)
      max = try container.decodeIfPresent(Int.self, forKey: .max)
      defaultValue = try container.decode(VariantSelection.Value.self, forKey: .defaultValue)
    }
  }

  public let id: String
  public let specVersion: Int
  public let rulesVersion: Int
  public let name: LocalizedText
  public let shortDescription: LocalizedText?
  public let symbol: String
  public let paletteID: PaletteToken
  public let players: Players
  public let scoring: Scoring
  public let engine: String
  public let end: End
  public let tieBreak: [TieBreakRule]
  public let variants: [Variant]
  public let statsProfiles: [String]

  private enum CodingKeys: String, CodingKey {
    case id, specVersion, rulesVersion, name, shortDescription, symbol, paletteID
    case players, scoring, engine, end, tieBreak, variants, statsProfiles
  }

  public init(
    id: String,
    specVersion: Int,
    rulesVersion: Int,
    name: LocalizedText,
    shortDescription: LocalizedText? = nil,
    symbol: String,
    paletteID: PaletteToken = .ink,
    players: Players,
    scoring: Scoring,
    engine: String,
    end: End,
    tieBreak: [TieBreakRule] = [.shared],
    variants: [Variant] = [],
    statsProfiles: [String] = ["standard"]
  ) {
    self.id = id
    self.specVersion = specVersion
    self.rulesVersion = rulesVersion
    self.name = name
    self.shortDescription = shortDescription
    self.symbol = symbol
    self.paletteID = paletteID
    self.players = players
    self.scoring = scoring
    self.engine = engine
    self.end = end
    self.tieBreak = tieBreak
    self.variants = variants
    self.statsProfiles = statsProfiles
  }

  public init(from decoder: Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)
    id = try container.decode(String.self, forKey: .id)
    specVersion = try container.decode(Int.self, forKey: .specVersion)
    rulesVersion = try container.decode(Int.self, forKey: .rulesVersion)
    name = try container.decode(LocalizedText.self, forKey: .name)
    shortDescription = try container.decodeIfPresent(LocalizedText.self, forKey: .shortDescription)
    symbol = try container.decode(String.self, forKey: .symbol)
    paletteID = try container.decodeIfPresent(PaletteToken.self, forKey: .paletteID) ?? .ink
    players = try container.decode(Players.self, forKey: .players)
    scoring = try container.decode(Scoring.self, forKey: .scoring)
    engine = try container.decode(String.self, forKey: .engine)
    end = try container.decode(End.self, forKey: .end)
    tieBreak = try container.decodeIfPresent([TieBreakRule].self, forKey: .tieBreak) ?? [.shared]
    variants = try container.decodeIfPresent([Variant].self, forKey: .variants) ?? []
    statsProfiles =
      try container.decodeIfPresent([String].self, forKey: .statsProfiles) ?? ["standard"]
  }

  /// Doc utilisateur — un seul endroit pour ce prédicat (Skyjo : qui a fermé la manche), repris
  /// à l'identique par `LiveMatchModel` et `SharedMatchModel` plutôt que dupliqué à chaque écran
  /// qui affiche la sélection.
  public var requiresCloserSelection: Bool {
    scoring.modifiers.contains { $0.kind == .exclusiveFlag && $0.required }
  }
}

public typealias Direction = GameDefinition.Direction
public typealias TieBreakRule = GameDefinition.TieBreakRule

File: android/domain/src/main/kotlin/com/quimene/domain/rules/GameDefinition.kt
package com.quimene.domain.rules

import com.quimene.domain.model.VariantSelection
import com.quimene.domain.model.VariantValue
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.util.Locale

/**
 * Miroir de `GameDefinition.swift` et de ses types imbriqués — un type Kotlin pour un type
 * Swift, mêmes noms, mêmes valeurs par défaut (posées ici comme valeurs de paramètre Kotlin,
 * plus simple que le `init(from:)` manuel de Swift). Champs et valeurs par défaut vérifiés
 * contre `spec/schema/game-definition.schema.json`, pas seulement contre le rapport
 * d'exploration du source Swift.
 */
@Serializable
data class GameDefinition(
    val id: String,
    val specVersion: Int,
    val rulesVersion: Int,
    val name: LocalizedText,
    val shortDescription: LocalizedText? = null,
    val symbol: String,
    val paletteID: PaletteToken = PaletteToken.Ink,
    val players: Players,
    val scoring: Scoring,
    val engine: String,
    val end: End,
    val tieBreak: List<TieBreakRule> = listOf(TieBreakRule.Shared),
    val variants: List<Variant> = emptyList(),
    val statsProfiles: List<String> = listOf("standard"),
) {
    val requiresCloserSelection: Boolean
        get() = scoring.modifiers.any { it.kind == ModifierKind.ExclusiveFlag && it.required }
}

/** Miroir de `GameDefinition.LocalizedText`. */
@Serializable
data class LocalizedText(
    val fr: String,
    val en: String? = null,
    val es: String? = null,
    val de: String? = null,
    val it: String? = null,
) {
    /** Langue système, repli sur [fr] — miroir de `localized` (`Bundle.main.preferredLocalizations`
     * côté Swift ; `Locale.getDefault()` en JVM pur, `:domain` n'a pas de contexte Android). */
    val localized: String
        get() {
            val translated =
                when (Locale.getDefault().language) {
                    "en" -> en
                    "es" -> es
                    "de" -> de
                    "it" -> it
                    else -> null
                }
            return translated ?: fr
        }

    /** Insensible aux espaces (pas seulement en début/fin — partout dans la chaîne, des deux
     * côtés) et à la langue affichée : compare contre les 5 traductions déclarées, pas seulement
     * celle actuellement montrée — miroir de `LocalizedText.matches(_:)` (Swift). */
    fun matches(term: String): Boolean {
        val needle = term.filter { !it.isWhitespace() }
        if (needle.isEmpty()) return false
        return listOfNotNull(fr, en, es, de, it).any { candidate ->
            candidate.filter { !it.isWhitespace() }.contains(needle, ignoreCase = true)
        }
    }
}

/** Miroir de `GameDefinition.PaletteToken`. */
@Serializable
enum class PaletteToken {
    @SerialName("ink")
    Ink,

    @SerialName("brass")
    Brass,

    @SerialName("teal")
    Teal,

    @SerialName("azur")
    Azur,

    @SerialName("ambre")
    Ambre,

    @SerialName("emeraude")
    Emeraude,

    @SerialName("magenta")
    Magenta,

    @SerialName("ardoise")
    Ardoise,

    @SerialName("cyan")
    Cyan,

    @SerialName("vermillon")
    Vermillon,

    @SerialName("violet")
    Violet,

    @SerialName("olive")
    Olive,
}

/** Miroir de `GameDefinition.Players`. */
@Serializable
data class Players(
    val min: Int,
    val max: Int,
    val recommended: List<Int> = emptyList(),
    val teams: Teams? = null,
) {
    /** Miroir de `GameDefinition.Players.Teams`. */
    @Serializable
    data class Teams(
        val size: Int,
        val fixed: Boolean = true,
    )
}

/** Miroir de `GameDefinition.EntryKind`. */
@Serializable
enum class EntryKind {
    @SerialName("integer")
    Integer,

    @SerialName("categorySheet")
    CategorySheet,

    @SerialName("structured")
    Structured,

    @SerialName("rank")
    Rank,

    @SerialName("predictionAndResult")
    PredictionAndResult,
}

/** Miroir de `GameDefinition.Category`. */
@Serializable
data class Category(
    val id: String,
    val label: LocalizedText,
    val section: Section,
    val scoring: Scoring,
) {
    @Serializable
    enum class Section {
        @SerialName("upper")
        Upper,

        @SerialName("lower")
        Lower,
    }

    /** Nom identique à `Scoring` top-niveau côté Swift (deux types distincts qui portent le même
     * nom court, désambiguïsés par l'imbrication — Kotlin le permet nativement, comme Swift). */
    @Serializable
    data class Scoring(
        val kind: Kind,
        val value: Int? = null,
        val max: Int? = null,
    ) {
        @Serializable
        enum class Kind {
            @SerialName("fixed")
            Fixed,

            @SerialName("sumOfDice")
            SumOfDice,

            @SerialName("multipleOf")
            MultipleOf,
        }
    }
}

/** Miroir de `GameDefinition.RankOption`. */
@Serializable
data class RankOption(
    val id: String,
    val label: LocalizedText,
    val points: Int,
)

/** Miroir de `GameDefinition.ScoreEntrySpec`. */
@Serializable
data class ScoreEntrySpec(
    val kind: EntryKind,
    val min: Int? = null,
    val max: Int? = null,
    val step: Int = 1,
    val allowsNegative: Boolean = false,
    val warnBelow: Int? = null,
    val warnAbove: Int? = null,
    val categories: List<Category>? = null,
    val ranks: List<RankOption>? = null,
)

/** Miroir de `GameDefinition.ModifierKind`. */
@Serializable
enum class ModifierKind {
    @SerialName("exclusiveFlag")
    ExclusiveFlag,

    @SerialName("flag")
    Flag,

    @SerialName("counter")
    Counter,
}

/**
 * Miroir de `GameDefinition.Modifier` — renommé `ScoreModifier` (pas `Modifier`) : `Modifier`
 * collisionnerait systématiquement avec `androidx.compose.ui.Modifier`, omniprésent dans le
 * code Compose des étapes suivantes. Mêmes champs, même sémantique, seul le nom diffère —
 * `Modifier` n'existe pas comme type au top-niveau côté Swift (pas de collision là-bas).
 */
@Serializable
data class ScoreModifier(
    val id: String,
    val label: LocalizedText,
    val kind: ModifierKind,
    val required: Boolean = false,
    val max: Int? = null,
)

/** Miroir de `GameDefinition.Direction`. */
@Serializable
enum class Direction {
    @SerialName("lowestWins")
    LowestWins,

    @SerialName("highestWins")
    HighestWins,

    @SerialName("targetExact")
    TargetExact,
}

/** Miroir de `GameDefinition.Aggregation`. */
@Serializable
enum class Aggregation {
    @SerialName("cumulative")
    Cumulative,

    @SerialName("lastRoundOnly")
    LastRoundOnly,

    @SerialName("bestRound")
    BestRound,
}

/** Miroir du `Scoring` top-niveau (`GameDefinition.swift`) — champ JSON `"modifiers"`, décodé en
 * `List<ScoreModifier>` (voir la note de renommage ci-dessus). */
@Serializable
data class Scoring(
    val direction: Direction,
    val aggregation: Aggregation = Aggregation.Cumulative,
    val roundLabel: LocalizedText? = null,
    val entry: ScoreEntrySpec,
    val modifiers: List<ScoreModifier> = emptyList(),
)

/** Miroir de `GameDefinition.EndConditionType`. */
@Serializable
enum class EndConditionType {
    @SerialName("scoreThreshold")
    ScoreThreshold,

    @SerialName("roundLimit")
    RoundLimit,

    @SerialName("targetReached")
    TargetReached,

    @SerialName("allSheetsComplete")
    AllSheetsComplete,

    @SerialName("elimination")
    Elimination,

    @SerialName("manualStop")
    ManualStop,
}

/** Miroir de `GameDefinition.Comparison` — valeurs JSON symboliques (`">="`, pas
 * `"greaterThanOrEqual"`), comme le raw value Swift. */
@Serializable
enum class Comparison {
    @SerialName(">=")
    GreaterThanOrEqual,

    @SerialName(">")
    GreaterThan,

    @SerialName("==")
    EqualTo,

    @SerialName("<=")
    LessThanOrEqual,

    @SerialName("<")
    LessThan,
    ;

    fun evaluate(
        lhs: Int,
        rhs: Int,
    ): Boolean =
        when (this) {
            GreaterThanOrEqual -> lhs >= rhs
            GreaterThan -> lhs > rhs
            EqualTo -> lhs == rhs
            LessThanOrEqual -> lhs <= rhs
            LessThan -> lhs < rhs
        }
}

/** Miroir de `GameDefinition.Scope`. */
@Serializable
enum class Scope {
    @SerialName("anyPlayer")
    AnyPlayer,

    @SerialName("allPlayers")
    AllPlayers,

    @SerialName("anyTeam")
    AnyTeam,
}

/** Miroir de `GameDefinition.EndCondition`. */
@Serializable
data class EndCondition(
    val type: EndConditionType,
    val value: Int? = null,
    val valueExpression: String? = null,
    val comparison: Comparison = Comparison.GreaterThanOrEqual,
    val scope: Scope = Scope.AnyPlayer,
    val completeRound: Boolean = true,
    val variantKey: String? = null,
) {
    fun resolvedValue(variants: VariantSelection): Int =
        if (variantKey != null) variants.int(variantKey, value ?: 0) else (value ?: 0)
}

/** Miroir de `GameDefinition.End`. */
@Serializable
data class End(
    val conditions: List<EndCondition>,
)

/** Miroir de `GameDefinition.TieBreakRule` / `TieBreakRule` top-niveau. */
@Serializable
enum class TieBreakRule {
    @SerialName("bestSingleRound")
    BestSingleRound,

    @SerialName("worstSingleRound")
    WorstSingleRound,

    @SerialName("fewestRoundsClosed")
    FewestRoundsClosed,

    @SerialName("mostRoundsWon")
    MostRoundsWon,

    @SerialName("lowerSecondaryScore")
    LowerSecondaryScore,

    @SerialName("higherSecondaryScore")
    HigherSecondaryScore,

    @SerialName("headToHead")
    HeadToHead,

    @SerialName("shared")
    Shared,
}

/** Miroir de `GameDefinition.VariantKind`. */
@Serializable
enum class VariantKind {
    @SerialName("bool")
    Bool,

    @SerialName("integerChoice")
    IntegerChoice,

    @SerialName("integerRange")
    IntegerRange,

    @SerialName("option")
    Option,
}

/** Miroir de `GameDefinition.Variant` — clé JSON `"default"` (mot réservé mou en Kotlin,
 * utilisable comme nom de propriété mais renommé `defaultValue` pour rester lisible en usage). */
@Serializable
data class Variant(
    val id: String,
    val label: LocalizedText,
    val help: LocalizedText? = null,
    val kind: VariantKind,
    val values: List<VariantValue> = emptyList(),
    val min: Int? = null,
    val max: Int? = null,
    @SerialName("default") val defaultValue: VariantValue,
)
